pkg/daemon/certificate_writer.go

```diff
@@ -0,0 +1,120 @@
+package daemon
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	"github.com/openshift/machine-config-operator/pkg/daemon/constants"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	ccRequeueDelay = 1 * time.Minute
+)
+
+func (dn *Daemon) handleControllerConfigEvent(obj interface{}) {
+	controllerConfig := obj.(*mcfgv1.ControllerConfig)
+	glog.V(4).Infof("Updating ControllerConfig %s", controllerConfig.Name)
+	dn.enqueueControllerConfig(controllerConfig)
+}
+
+func (dn *Daemon) enqueueControllerConfig(controllerConfig *mcfgv1.ControllerConfig) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(controllerConfig)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %w", controllerConfig, err))
+		return
+	}
+	dn.ccQueue.AddRateLimited(key)
+}
+
+func (dn *Daemon) enqueueControllerConfigAfter(controllerConfig *mcfgv1.ControllerConfig, after time.Duration) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(controllerConfig)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %w", controllerConfig, err))
+		return
+	}
+
+	dn.ccQueue.AddAfter(key, after)
+}
+
+func (dn *Daemon) controllerConfigWorker() {
+	for dn.processNextControllerConfigWorkItem() {
+	}
+}
+
+func (dn *Daemon) processNextControllerConfigWorkItem() bool {
+	key, quit := dn.ccQueue.Get()
+	if quit {
+		return false
+	}
+	defer dn.ccQueue.Done(key)
+
+	err := dn.syncControllerConfigHandler(key.(string))
+	dn.handleControllerConfigErr(err, key)
+
+	return true
+}
+
+func (dn *Daemon) handleControllerConfigErr(err error, key interface{}) {
+	if err == nil {
+		dn.ccQueue.Forget(key)
+		return
+	}
+
+	if err := dn.updateErrorState(err); err != nil {
+		glog.Errorf("Could not update annotation: %v", err)
+	}
+	// This is at V(2) since the updateErrorState() call above ends up logging too
+	glog.V(2).Infof("Error syncing ControllerConfig %v (retries %d): %v", key, dn.ccQueue.NumRequeues(key), err)
+	dn.ccQueue.AddRateLimited(key)
+}
+
+func (dn *Daemon) syncControllerConfigHandler(key string) error {
```

Comment:
question: Will this run outside of the usual `dn.update()` process? Also, could it potentially write different content to `caBundleFilePath` than the MachineConfig contains?

I ask these questions because of Config Drift Monitor. If the file is part of the MachineConfig, Config Drift Monitor will expect its contents to match. It sounds like at a minimum, Config Drift Monitor should ignore `caBundleFilePath` since we're managing it here.

Reply:
Right, i explicitly added 2 clauses in validation and writes so this would happen completely separately. This does of course cause 2 issues implicitly:

1. the config drift monitor keeps triggering whenever the file changes, see nothing changed, might be confused, but then lets it carry on
2. the node still says it updates (rebootless), performs the update, but doesn't touch the file (if its the only change), so it technically does a fake update

I can try to make these processes better but they don't block anything I believe